src/main/java/edu/kit/compiler/optimizations/ArithmeticIdentitiesOptimization.java

```diff
@@ -0,0 +1,443 @@
+                        exchange(node, enqueued(subNode));
+                    }
+                }
+                case iro_Sub -> {
+                    // -(x - y) --> y - x
+                    assert operand.getPredCount() == 2;
+                    var left = operand.getPred(0);
+                    var right = operand.getPred(1);
+                    var sub = graph.newSub(node.getBlock(), right, left);
+                    exchange(node, enqueued(sub));
+                }
+                case iro_Mul -> {
+                    assert operand.getPredCount() == 2;
+                    var constValue = getConstValue(operand.getPred(1));
+                    if (constValue != null) {
+                        // - (x * c) --> x * -c
+                        var constNode = graph.newConst(constValue.neg());
+                        var mulNode = graph.newMul(node.getBlock(), operand.getPred(0), constNode);
+                        exchange(node, enqueued(mulNode));
+                    }
+                }
+                default -> {
+                }
+            }
+        }
+
+        @Override
+        public void visit(Mul node) {
+            if (isConstOne(node.getLeft())) {
+                // 1 * x --> x
+                exchange(node, node.getRight());
+            } else if (isConstOne(node.getRight())) {
+                // x * 1 --> x
+                exchange(node, node.getLeft());
+
+            } else if (isConstZero(node.getLeft())) {
+                // 0 * x --> 0
+                var zero = node.getRight().getMode().getNull();
+                exchange(node, graph.newConst(zero));
+            } else if (isConstZero(node.getRight())) {
+                // x * 0 --> 0
+                var zero = node.getLeft().getMode().getNull();
+                exchange(node, graph.newConst(zero));
+
+            } else if (isConstNegOne(node.getLeft())) {
+                // -1 * x --> -x
+                var minus = graph.newMinus(node.getBlock(), node.getRight());
+                exchange(node, enqueued(minus));
+            } else if (isConstNegOne(node.getRight())) {
+                // x * -1 --> -x
+                var minus = graph.newMinus(node.getBlock(), node.getLeft());
+                exchange(node, enqueued(minus));
+
+            } else if (isOnlyLeftConst(node.getLeft(), node.getRight())) {
+                // Normalization: c * x --> x * c
+                var mul = graph.newMul(node.getBlock(), node.getRight(), node.getLeft());
+                exchange(node, enqueued(mul));
+
+            } else if (node.getRight().getOpCode() == ir_opcode.iro_Const) {
+                // (x * c) * d --> x * (c * d)
+                foldMultiplication(node);
+            }
+        }
+
+        @Override
+        public void visit(Div node) {
+            if (isConstOne(node.getRight())) {
+                // x / 1 --> x
+                exchangeDivOrMod(node, node.getMem(), node.getLeft());
```

Comment:
Is it possible that `node` and `node.getMem()` are in different blocks? If so, would we get problems when exchanging the memory successor of the node with its memory predecessor?

Reply:
I don't think so, `exchange` moves all references from the successor `Mem` to the predecessor. I've tested it with the following code and it the optimization works as expected.

```java
public int div(int x, int y) {
    if (x < 10) {
        return x / 1;
    } else {
        return y / 1;
    }
}
```